lib/media-controls-impl.js

```diff
@@ -0,0 +1,524 @@
+;(function(define){'use strict';define(function(require,exports,module){
+
+/*
+** MediaControlsImpl object
+*/
+function MediaControlsImpl(mediaControlsElement, shadowRoot, player) {
+  this.mediaControlsElement = mediaControlsElement;
+  this.shadowRoot = shadowRoot;
+  this.touchStartID = null;
+  this.isPausedWhileDragging = null;
+  this.dragging = false;
+  this.sliderRect = null;
+  this.playedUntilEnd = false;
+  this.intervalId = null;
+  this.pausedAtEndWhilePlaying = false;
+  this.mediaPlayer = player;
+  this.seekIncrement = 10; // Seek forward/backward in 10 second increments
+  this.mouseEventHandlerRegistered = false;
+
+  this.els = {
+    durationText: this.shadowRoot.querySelector('.duration-text'),
+    elapsedText: this.shadowRoot.querySelector('.elapsed-text'),
+    elapsedTime: this.shadowRoot.querySelector('.elapsed-time'),
+    play: this.shadowRoot.querySelector('.play'),
+    playHead: this.shadowRoot.querySelector('.play-head'),
+    seekForward: this.shadowRoot.querySelector('.seek-forward'),
+    seekBackward: this.shadowRoot.querySelector('.seek-backward'),
+    sliderWrapper: this.shadowRoot.querySelector('.slider-wrapper')
+  };
+
+  this.addEventListeners();
+}
+
+MediaControlsImpl.prototype.addEventListeners = function() {
+  this.shadowRoot.addEventListener('touchstart', this);
+  this.shadowRoot.addEventListener('touchmove', this);
+  this.shadowRoot.addEventListener('touchend', this);
+  this.shadowRoot.addEventListener('mousedown', this);
+
+  this.mediaPlayer.addEventListener('loadedmetadata', this);
+  this.mediaPlayer.addEventListener('play', this);
+  this.mediaPlayer.addEventListener('pause', this);
+  this.mediaPlayer.addEventListener('timeupdate', this)
+  this.mediaPlayer.addEventListener('seeked', this);
+  this.mediaPlayer.addEventListener('ended', this);
+};
+
+MediaControlsImpl.prototype.removeEventListeners = function() {
+  this.shadowRoot.removeEventListener('touchstart', this);
+  this.shadowRoot.removeEventListener('touchmove', this);
+  this.shadowRoot.removeEventListener('touchend', this);
+  this.shadowRoot.removeEventListener('mousedown', this);
+
+  this.mediaPlayer.removeEventListener('loadedmetadata', this);
+  this.mediaPlayer.removeEventListener('play', this);
+  this.mediaPlayer.removeEventListener('pause', this);
+  this.mediaPlayer.removeEventListener('timeupdate', this)
+  this.mediaPlayer.removeEventListener('seeked', this);
+  this.mediaPlayer.removeEventListener('ended', this);
+
+  if (this.mouseEventHandlerRegistered) {
+    this.shadowRoot.removeEventListener('mousemove', this, true);
+    this.shadowRoot.removeEventListener('mouseup', this, true);
+  }
+
+  this.mouseEventHandlerRegistered = false;
+};
+
+MediaControlsImpl.prototype.handleEvent = function(e) {
+
+  // If we get a touchstart, don't process mouse events
+  if (e.type === 'touchstart') {
+    this.shadowRoot.removeEventListener('mousedown', this);
+    this.shadowRoot.removeEventListener('mousemove', this);
```

Comment:
For mousemove and mouseup, you need the third argument true to really deregister them.

Actually, the way this is written, you're going to be making these threee removeEventLIstener() calls on every single touchstart event forever.  You should probably have a flag so you only need to do it the first time.